♻️ [bento][amp-social-share] Add prop types and update type definitions for Social Share and basic cleanup

krdwan opened the pull request:
- Added prop types
- Updated type definitions throughout
- Used `obj.key` operator instead of `obj['key']` when possible

dvoytenko commented on extensions/amp-social-share/1.0/social-share-config.js:
```diff
@@ -14,19 +14,23 @@
 /**
- * @typedef {!Object<string, !JsonObject|string>}
+ * @typedef {{
+ *   shareEndpont:           string,
```
Nit: By the styleguide we just need one space between `:` and type.

krdwan replied:
Thanks, updated throughout all files!

dvoytenko commented on extensions/amp-social-share/1.0/social-share-config.js:
```diff
@@ -14,19 +14,23 @@
+ *   defaultParams:          Object<string, string>,
+ *   defaultColor:           string,
+ *   defaultBackgroundColor: string,
+ *   bindings:               ?Array<string>,
```
This should actually be `(!Array<string>|undefined)`. I.e. it's either absent or has a real value. The `?` indicates `nullable`, but it's not the case here.

krdwan replied:
Thanks, I've done a full audit and removed most `?` throughout the files.  I had confused `nullable` and `|undefined`.

dvoytenko commented on extensions/amp-social-share/1.0/social-share-config.js:
```diff
@@ -36,99 +40,104 @@ let SocialShareConfigDef;
  *   {string} shareEndpoint - The base API endpoint for sharing to the
  *     specified social media type.
- *   {JsonObject} defaultParams - Parameters to be appended to the end of the
+ *   {Object} defaultParams - Parameters to be appended to the end of the
```
Let's move this per-property description to the `SocialShareConfigDef` type.

krdwan replied:
Thanks, updated!

dvoytenko commented on extensions/amp-social-share/1.0/social-share-config.js:
```diff
@@ -36,99 +40,104 @@ let SocialShareConfigDef;
+ *     social media type.
+ *   {?Array<string>} bindings - Used for email, allows passing in an
+ *     attribute that be used in endpoint, but not as a search param
+ * @type {Object<?string, SocialShareConfigDef>}
```
`null` key is not allowed.

krdwan replied:
Thank you, updated!

dvoytenko commented on extensions/amp-social-share/1.0/social-share-config.js:
```diff
@@ -36,99 +40,104 @@ let SocialShareConfigDef;
-    'defaultBackgroundColor': '1da1f2',
+    shareEndpoint: 'https://twitter.com/intent/tweet',
+    defaultParams: {
+      text: 'TITLE',
```
Because the `defaultParams` is defined as `!Object`, we should also use `dict()` and quotes. You can think of it this way: you don't define `text` and `url` fields anywhere - they are simply names. Some other name could be "this is the text" and it's still a reasonably-allowable name, even if unlikely.

krdwan replied:
Ah thanks, so to confirm anytime the keys of the object would be the value of a variable (not the name of the variable), we should use `dict()` and quotes for that object? 

I've updated in the latest revision.

dvoytenko commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -24,33 +24,36 @@ import {getSocialConfig} from './social-share-config';
-const DEFAULT_HEIGHT = 44;
-const DEFAULT_TARGET = '_blank';
 const NAME = 'SocialShare';
+const DEFAULT_WIDTH = '60';
```
Why do these need to become strings?

krdwan replied:
Initially I had wanted to set the `width` and `height` props to `(string|undefined)`, so wanted to follow that format throughout the code.

In the latest, I decided to change it to `(number|string|undefined)` so this is reflected in the latest revision.  Thanks!!

dvoytenko commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -76,27 +79,22 @@ export function SocialShare(props) {
  * values.
- * @param {!JsonObject} props
- * @param {JsonObject} size
+ * @param  {?string}               type
```
Ditto with the namespace.

krdwan replied:
Thanks, updated!

dvoytenko commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -108,33 +106,29 @@ function processChildren(props, size) {
 
 /**
- * @param {!JsonObject} props
+ * @param {?string}                     type
```
ditto

krdwan replied:
Thanks, updated!

dvoytenko commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -233,9 +227,9 @@ function isIos() {
 
 /**
- * @param {!Event} event
+ * @param {!Event}  event
```
Ditto: namespace

krdwan replied:
Thanks, updated!

dvoytenko commented on extensions/amp-social-share/1.0/social-share.type.js:
```diff
@@ -0,0 +1,34 @@
+
+/**
+ * @typedef {{
+ *   type:       ?string,
```
Ditto: namespace.

krdwan replied:
Thanks, updated!

dvoytenko commented on extensions/amp-social-share/1.0/social-share.type.js:
```diff
@@ -0,0 +1,34 @@
+/**
+ * @typedef {{
+ *   type:       ?string,
+ *   endpoint:   ?string,
```
I think all of these are actually `type|undefined` and not `?type`. In other words - this prop can be omitted, but when specified, the valid value should be given. And `null` typically is not valid.

krdwan replied:
Thanks, updated in the latest revision.  I had confused `nullable` and `|undefined`.

caroqliu commented on extensions/amp-social-share/1.0/social-share-config.js:
```diff
@@ -14,19 +14,23 @@
- * @param  {string} type
- * @return {!SocialShareConfigDef}
+ * @param  {?string} type
+ * @return {?SocialShareConfigDef}
```
`SocialShareConfigDef|undefined`?

caroqliu replied:
Alternatively we can `return BUILTINS[type] || {}` and remove the `|| {}` from where this is called in `SocialShare`

krdwan replied:
Thanks, updated!

caroqliu commented on extensions/amp-social-share/1.0/social-share-config.js:
```diff
@@ -14,19 +14,23 @@
+ *   bindings:               ?Array<string>,
+ * }}
  */
 let SocialShareConfigDef;
```
WDYT about moving this to `social-share.type.js` so it can share the namespace?

krdwan replied:
I like the idea of defining more types in `social-share.type.js`.  I also like the idea of keeping it in the config file since all the data is in the config so easy to keep everything together.

I'm not sure I fully understand the implications regarding namespace, could you elaborate?

caroqliu replied:
If we put it all in the `SocialShareDef` namespace, this can be `SocialShareDef.Config`, much like we do for other Bento components i.e. `DateDisplay`. It's just a rec for consistency is all.

krdwan replied:
Sure, updated in latest version.

caroqliu commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -24,33 +24,36 @@ import {getSocialConfig} from './social-share-config';
+  background,
+  tabIndex,
+  style,
+  children,
```
Let's also propagate `...rest`

krdwan replied:
Thanks, I see this is a common pattern for our other components.  For this specific instance, I'm not exactly sure where we would want to propagate it too.  For the most part, we don't really want the icon to change that much except for colors that we allow, nor do we want to change the wrapper `div` that much.  If there's something I'm missing, could you elaborate?

caroqliu replied:
There could be a number of attributes we want to support generically, i.e. `aria-label`, `aria-disabled`, `disabled` -- if this is a button I assume we want to respect if it's not clickable? I don't think we should try to support these exhaustively, hence `rest`. If there are ones we explicitly do not want to support then that is worth discussing too.

krdwan replied:
Sounds good, added!

caroqliu commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -60,43 +63,36 @@ export function SocialShare(props) {
     <div
       role="button"
-      tabindex={props['tabIndex'] || '0'}
+      tabindex={tabIndex || '0'}
```
Let's give `tabIndex` its default value when you destructure it instead of here.

krdwan replied:
Thanks, updated!

caroqliu commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -24,33 +24,36 @@ import {getSocialConfig} from './social-share-config';
     checkedHeight,
     checkedTarget,
-  } = checkProps(props);
+  } = checkProps(type, endpoint, target, width, height, params);
```
You should still be able to pass `props` -- would be easier if you ever need to `checkProps` on anything new without having to change the fn signature.

caroqliu replied:
Note that `DateDisplay` is an example of a component that passes `props` to helper functions.

krdwan replied:
Thanks!  Right we definitely can still use props.  One thing, I was thinking is that we do lose all the type definitions for these internal functions if we change to props.  I'm don't feel strongly about it either way, but I am thinking we can use this PR to clearly define all types within Social Share.

I also noticed that some of our bento components do not use this `prop` pattern such as `timeago`, `fit-text`.  Thoughts?

caroqliu replied:
Why would we lose the type definitions if we change to props? Won't they still be defined as properties of the prop definition provided?

caroqliu replied:
IMO `checkProps` in `SocialShare` is ostensibly more generic than `getFuzzyTimestampValue` whose inputs are unlikely to change. But this is up to you.

krdwan replied:
Sounds good, since I just defined all the types for `checkProps` and `processChildren` functions in this PR, I'd prefer not to call them with `props` :)

caroqliu commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -60,43 +63,36 @@ export function SocialShare(props) {
+      style={{...size, ...style}}
     >
-      {processChildren(props, size)}
+      {processChildren(type, children, color, background, size)}
```
Same as above, let's keep this passing `props`

caroqliu commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -60,43 +63,36 @@ export function SocialShare(props) {
     return children;
   } else {
-    const typeConfig = getSocialConfig(type) || dict();
+    const typeConfig = getSocialConfig(/** @type {string} */ (type)) || {};
```
If `type` is expected to be a `string`, why not remove `undefined` here and in the type def? That way you won't have to cast it here.

caroqliu replied:
Alternatively if it is not expected to be a string, you could give it the empty string as a default value.

krdwan replied:
Yep, so in the `checkProps` function, the `type` will be verified to be a string or it will error out.  So at this point, it can only be a string.  But, since we are still using the original `type` variable with the typedef of `(string|undefined)`, I have to use the cast.  That's my understanding, maybe I am missing your point?

caroqliu replied:
If `type` is `string` by the time `processChildren` is called we could update the signature for that at least?

krdwan replied:
Sure, I updated the function signature, and casted type before calling the function in the latest version.

krdwan commented:
Thanks for review @dvoytenko 

Latest revision includes:

- Address requested changes
- Make width and height `number|string`
- Added / editted a few existing comments

One thing, I was considering is that I tended to stay away from `!` operator for types.  i.e. I would go with `string` over `!string` in most cases as they seem to be the same to me, i.e.

- nullable string: ?string
- non-nullable string: string or !string

Let me know if you think I should change

krdwan commented:
Thanks for review @caroqliu 

I updated the tabIndex default value in the latest update.  Responses to your comments are in line!  Thanks!

krdwan commented:
Thanks @jridgewell @alanorozco for the review.

@caroqliu Latest version includes the following-

1. Moved config typedef into social-share.types.js
2. Move casting `type` to string, to outside of `processChildren` and updated the function signature.

Other reponses to your comments are in line.
